File: lib/domain/field/ports/field_update_output_port.rb
# frozen_string_literal: true

module Domain
  module Field
    module Ports
      class FieldUpdateOutputPort
        def on_success(field)
          raise NotImplementedError, "Subclasses must implement on_success"
        end

        def on_failure(error_dto)
          raise NotImplementedError, "Subclasses must implement on_failure"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/field/ports/field_update_output_port.rs
use crate::field::entities::FieldEntity;
use crate::shared::dtos::error::Error;
use crate::shared::policies::policy_permission_denied::PolicyPermissionDenied;

/// Ruby: `Domain::Field::Ports::FieldUpdateOutputPort`
pub trait FieldUpdateOutputPort {
    fn on_success(&mut self, field: FieldEntity);
    fn on_failure(&mut self, error: UpdateFailure);
}

#[derive(Debug, Clone, PartialEq)]
pub enum UpdateFailure {
    Policy(PolicyPermissionDenied),
    Error(Error),
}
